src/libsyntax/ext/placeholders.rs

```diff
@@ -32,12 +32,12 @@ pub fn placeholder(kind: AstFragmentKind, id: ast::NodeId) -> AstFragment {
         attrs: ThinVec::new(),
         kind: ast::ExprKind::Mac(mac_placeholder()),
     });
-    let ty = P(ast::Ty {
+    let ty = || P(ast::Ty {
         id,
         kind: ast::TyKind::Mac(mac_placeholder()),
         span,
     });
-    let pat = P(ast::Pat {
+    let pat = || P(ast::Pat {
```

Comment:
Just for learning, what's the benefit of this to-closure diff?

Reply:
In the old code, `pat` is only used sometimes. But we always pay the cost of constructing it. By using a closure, the construction cost is only paid when the closure is called.

It's similar to the difference between [`Option::unwrap_or`](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or) and `Option::unwrap_or_else`.